src/Chainweb/Pact/TransactionExec.hs

```diff
@@ -94,66 +97,62 @@ applyCmd
     -> SPVSupport
     -> Command (Payload PublicMeta ParsedCode)
     -> IO (CommandResult [TxLog Value])
-applyCmd logger pactDbEnv minerInfo gasModel pd spv cmd = do
+applyCmd l pactDbEnv minerInfo gasModel pd0 spv cmd = do
 
     let userGasEnv = mkGasEnvOf cmd gasModel
-        requestKey = cmdToRequestKey cmd
-        pd' = set pdPublicMeta (publicMetaOf cmd) $ pd
+        pd = set pdPublicMeta (publicMetaOf cmd) $ pd0
         supply = gasSupplyOf cmd
 
-    let buyGasEnv = CommandEnv Nothing Transactional pactDbEnv logger freeGasEnv pd'
-
-    buyGasResultE <- catchesPactError $! buyGas buyGasEnv cmd spv minerInfo supply
-
-    case buyGasResultE of
-
-      Left e1 ->
-        jsonErrorResult buyGasEnv requestKey e1 [] (Gas 0)
-          "tx failure for requestKey when buying gas"
-
-      Right buyGasResult -> do
-
-        -- this call needs to fail hard if Left. It means the continuation did not process
-        -- correctly, and we should fail the transaction
-        (!pactId, !buyGasLogs) <- either internalError pure buyGasResult
-        logDebugRequestKey logger requestKey "successful gas buy for request key"
-
-        let !payloadEnv = set ceGasEnv userGasEnv
-              $ set cePublicData pd' buyGasEnv
-
-        cmdResultE <- catchesPactError $! runPayload payloadEnv def cmd spv buyGasLogs
-
-        case cmdResultE of
-
-          Left e2 ->
-
-            jsonErrorResult payloadEnv requestKey e2 buyGasLogs (Gas 0)
-              "tx failure for request key when running cmd"
-
-          (Right !cmdResult) -> do
+    let buyGasEnv = CommandEnv Nothing Transactional pactDbEnv l freeGasEnv pd spv
 
-            logDebugRequestKey logger requestKey "success for requestKey"
-
-            let !redeemGasEnv = set ceGasEnv freeGasEnv payloadEnv
-                cmdLogs = fromMaybe [] $ _crLogs cmdResult
-            redeemResultE <- catchesPactError $!
-              redeemGas redeemGasEnv cmd cmdResult pactId supply spv cmdLogs
-
-            case redeemResultE of
-
-              Left e3 ->
-
-                jsonErrorResult redeemGasEnv requestKey e3 cmdLogs (_crGas cmdResult)
-                  "tx failure for request key while redeeming gas"
-
-              (Right !redeemResult) -> do
-
-                let !redeemLogs = fromMaybe [] $ _crLogs redeemResult
-                    !finalResult = over (crLogs . _Just) (<> redeemLogs) cmdResult
-
-                logDebugRequestKey logger requestKey "successful gas redemption for request key"
-                pure finalResult
+    withGas l buyGasEnv cmd minerInfo supply userGasEnv $ \cenv cmd0 logs ->
+      runPayload cenv def cmd0 logs
 
+-- | A handle for wrapping 'runPayload' commands in a buy/redeem gas phase.
+--
+withGas
```

Comment:
This is fine but it didn't buy us much in terms of clarity ... you just moved 95% of the really confusing code from `applyCmd` to `withGas`

Reply:
Discussed - I'll refactor `applyCmd` to be clearer about control flow instead of moving to `withGas`